### PR TITLE
Refactor: career form

### LINKED ISSUE #503: Forms: Technical Debt
Let's remove the technical debt from our forms environment. We have already started collecting the individual form components in a folder together with storybook stories.

+ Remove the legacy components in `components/legacy/form` or integrate them in our `components/forms` folder properly
+ Align the technical structure of `Career Form` and `Contact Form` so we don't have two different approaches.
   - Ensure `useController` 
+ If anything is business crucial it might be the forms so we want to write tests for them:
  - Test if errors are shown
  - Test if submission is triggered
+ Provide a storybook story
+ Make sure no `legacy` component is used, and if it's used make sure you either use the updated counterpart of update the legacy component.

This is a wide issue, involves a lot of refactoring and decision which makes it a bade issue for starters.

### PR DESCRIPTION
**Pull request by milljoniaer**
### What is included? 
- refactored career form with `useController` (expect FileUpload) -> just like the contact form
- use of the new design form fields

### Thoughts on the PR
I tried to develop the career form just like Felix Hofmann the contact form. That is why I used the controlled approach. But in Addition to the Textfields and Checkbox, I had to deal with the File Dropper, which uses "react-dropzone".
I was not able to get "react-dropzone" to work together with the short and clean controlled approach of "react-hook-form" and had to use the manual approach for the File Dropper which  requires more code and even some ugly workarounds in the Form component to make the File Upload required and to validate the File Categories.

Because of the size of the PR, I will provide some thoughts of mine about some of the changed components.

#### Checkbox
--- Update ---
handled in #511 

#### File Dropper
--- Update ---
handled in #512 

#### Upload
I copied most of the code from the previous upload, but I had to change the way documents are uploaded, since the categories are passed differently than before by the new form.

### Example
[Staff Engineer](https://satellytescommain21751-chorehoverstateformcontact.gtsb.io/career/staff-software-engineer/)

### Screenshots
Normal:
image

Error:
image

Success:
image

### Questions:
Do we need an Uploading State? e.g. a loading bar
There is nothing like this defined in the design.

as far as i can see, it closes #503

### CODE REVIEW

**Comment by ghost** [comment]
# [Gatsby Cloud](https://www.gatsbyjs.com) Build Report
**satellytes.com-main**

:tada: Your build was successful! [See the Deploy preview here.](https://satellytescommain21751-chorehoverstateformcontact.gtsb.io)

## Build Details

[View the build logs here.](https://gatsbyjs.com/dashboard/f215130d-15e3-4176-9252-e13166fe48de/sites/674be368-adf0-4ac5-8fae-f9a4493adc51/builds/637f00a8-1718-48b8-9f32-fbeb45cfbad2/details)

:clock1: Build time: **1m**

## Performance

### Lighthouse report

| Metric | Score |
| ----- | ---- |
| Performance |   :large_orange_diamond: 78   |
| Accessibility |   :green_heart: 97   |
| Best Practices |   :green_heart: 100   |
| SEO |   :large_orange_diamond: 79   |

:link: [View full report](https://build-637f00a8-1718-48b8-9f32-fbeb45cfbad2.gtsb.io/reports/lighthouse/index.html)

**Comment by georgiee** [comment]
Can you give me an update about this draft. If you did not work on that I suggest to close the draft (the branch stays there) and come back to a PR once you are ready to work on it again. Thanks.

**Comment by milljoniaer** [comment]
Yes, I had some troubles with this issue, but I think I am "PR-ready" again.

**Comment by georgiee** [comment]
I added Fabian, Felix and me to the list of explicit reviewers. It's a rather large PR and the topic of refactoring and go in many directions. Let's support @milljoniaer who provided this piece of change to land it in main soon 💪

File: src/components/forms/checkbox/checkbox.tsx
Comment (georgiee) [inline src/components/forms/checkbox/checkbox.tsx]: + The checkbox is missed when I tab through the form
+ There is no focus highlight. We have a yellow border when interacting with the keyboard. We need that to signal where the focus sits right now
+ Is it feasible to deliver only this checkbox first? That would greatly simplify the review process as you could do the same for the other components too.

File: src/components/forms/file-dropper/file-list-item.tsx
Comment (georgiee) [inline src/components/forms/file-dropper/file-list-item.tsx]: That's the change required to make it translatable isn't it ?

File: src/components/forms/text-area/text-area.tsx
Comment (georgiee) [inline src/components/forms/text-area/text-area.tsx:21]: This is a design change. Can you help me where this is coming from? I remember something mentioned by eric before. What's the purpose of this if the highlight is not maintained. There is not a single highlight for when I'm inside a field. Is this an error or by design? Because I would reject that design then, it's not accessible. Can you help me?

File: src/components/pages/blog/not-avilable-in-german.tsx
Comment (georgiee) [inline src/components/pages/blog/not-avilable-in-german.tsx]: This is not related to the form but you got here by fixing the send button? Do we have other unrelated changes like that? You could bundle them and deliver them as an isolated refactoring.

File: src/components/pages/career-details/new-career-form/career-form.tsx
Comment (georgiee) [inline src/components/pages/career-details/new-career-form/career-form.tsx:44]: This will is still a monster. Can you think of ways to divide and conquer? Like putting the request with axios into a distinct file. Or creating smaller components like `CareerFileUpload` that just wraps the following:

```
<FileDropper
          register={register}
          setValue={setValue}
          setError={setError}
          clearErrors={clearErrors}
          errors={errors}
          name="documents"
          fileCategories={fileCategories}
          acceptedFileTypes={'.pdf'}
          illustration="monitor_024"
          maxFiles={3}
          validator={validator}
        />
```

That way you could have a very semantic form described while you push away the details of the implementation into a separate component that you could even collect in a separate file like `form-components`. That's an idea. Maybe it doesn't make sense in the end but we have to do something about the size because I don't feel comfortable looking at it and I don't think people feel comfortable touching and extending it in the future.

```
<!-- other fields -->
<ApplicationFileUploader/>

```

File: src/components/pages/career-details/new-career-form/career-form.tsx
Comment (georgiee) [inline src/components/pages/career-details/new-career-form/career-form.tsx]: This could also be a local component `CareerFormErrors` or such.

**Review by georgiee — CHANGES_REQUESTED** [review]
1. Deploy of netlify storybook failed, I would like to check, do you know what's wrong?
2. All that deleted stuff, I like 👍
3. I provided some general comments.

I think we have to work towards a few things before merging it but I see that we are going into a great direction already 💪

File: src/components/forms/text-area/text-area.tsx
Comment (milljoniaer) [reply src/components/forms/text-area/text-area.tsx:21]: This a design change, which was mentioned I think by Eric. The other form components act like this too

File: src/components/forms/file-dropper/file-list-item.tsx
Comment (milljoniaer) [reply src/components/forms/file-dropper/file-list-item.tsx]: yes

File: src/components/forms/text-area/text-area.tsx
Comment (georgiee) [reply src/components/forms/text-area/text-area.tsx:21]: Ok. We leave it as is. But we have to come back to this because it's not good for the user. 
Edit: In addition I see it's not good right now anyway with the focus. We need to accommodate for accessible. Please keep that in mind in general.

File: src/components/pages/blog/not-avilable-in-german.tsx
Comment (milljoniaer) [reply src/components/pages/blog/not-avilable-in-german.tsx]: That was something I had to do when I was deleting the legacy form fields. The former used button was defined in the forms

File: src/components/pages/career-details/new-career-form/career-form.tsx
Comment (milljoniaer) [reply src/components/pages/career-details/new-career-form/career-form.tsx]: 👍

File: src/components/pages/blog/not-avilable-in-german.tsx
Comment (milljoniaer) [reply src/components/pages/blog/not-avilable-in-german.tsx]: I sorted them out to another PR, should not be anymore in this one 👍

File: src/components/pages/career-details/new-career-form/career-form-fields.tsx
Comment (georgiee) [inline src/components/pages/career-details/new-career-form/career-form-fields.tsx:142]: Thanks for that refactoring

File: src/components/pages/career-details/new-career-form/career-form.tsx
Comment (georgiee) [reply src/components/pages/career-details/new-career-form/career-form.tsx]: seen that, thanks!

File: src/components/pages/career-details/new-career-form/career-form.tsx
Comment (georgiee) [reply src/components/pages/career-details/new-career-form/career-form.tsx:44]: I saw the new field `CareerDetailsFileUpload` 🙌

File: src/components/pages/career-details/career-details.stories.mdx
Comment (georgiee) [inline src/components/pages/career-details/career-details.stories.mdx]: Something is off here (via https://deploy-preview-398--satellytes-website-storybook.netlify.app/?path=%2Fdocs%2Fcomponents-pages-career-details--job-description)

The args table at the end is empty and there is an additional source code block printed before the section about `CareerForm` while there is one `Form` source block at the end without any paramters.

Screenshot 2022-03-28 at 11 53 18

File: src/components/pages/career-details/new-career-form/career-form.tsx
Comment (georgiee) [inline src/components/pages/career-details/new-career-form/career-form.tsx:17]: Thanks for putting some stuff aside that helps reading the form 🙌

File: src/components/pages/career-details/new-career-form/upload.ts
Comment (georgiee) [inline src/components/pages/career-details/new-career-form/upload.ts]: This is not only uploading something but it's the from submission. I'm thankful that you manage to lock away the logic for this in a separate file but the naming must be better. `submitApplication` for example tells the whole story.

Comment (georgiee) [reply src/components/pages/career-details/new-career-form/upload.ts]: This is by the way some crucial part of the form submission and always worth a test. Make sure you come back to have some fun with mocking the http layer and testing the expected outputs regarding errors messages, data submitted etc.

File: src/components/pages/career-details/new-career-form/career-form.tsx
Comment (georgiee) [inline src/components/pages/career-details/new-career-form/career-form.tsx:84]: You could place this is in a separate file like the submit logic which you place in `upload` already. You could easily test the logic here then too (next PR).  That's by the way nothing mandatory only if you like it. Refactoring is for me something to try out, then look at the new state and work with it. So it's super hard to describe or have suggestions without changing my mind every day 😄

File: src/components/pages/career-details/new-career-form/career-form-fields.tsx
Comment (georgiee) [inline src/components/pages/career-details/new-career-form/career-form-fields.tsx:75]: much better grouped like that instead of defining it inline in the form render part, or how does it fell to you ?

**Review by georgiee — APPROVED** [review]
Hey @milljoniaer thank you for that refactoring awesomeness and all of your efforts 🙏 I think we are indeed done!

I went through all changes but did not focus on small code nitpicks but rather the bigger picture to come to an end of this long running PR.

I left a few more comments but nothing that would block this PR anymore. 
You/We have already tested the form but things have changed in the mean time. Maybe you could test a last time if everything works end to end ? Are you also sure that the contact form is not affected by any of the changes in the previous small PRs and this last PR of the refactoring? Or should we test that too? I'm lacking a little bit the confidence if everything is still good, I also take your word instead of the testing 🙌

Let's merge when you are ready & confident enough.
Thank you!

File: src/components/pages/career-details/career-details.stories.mdx
Comment (milljoniaer) [reply src/components/pages/career-details/career-details.stories.mdx]: Oohps some leftovers of the old form... 🧹

File: src/components/pages/career-details/new-career-form/career-form.tsx
Comment (milljoniaer) [reply src/components/pages/career-details/new-career-form/career-form.tsx:17]: 👍

File: src/components/pages/career-details/new-career-form/upload.ts
Comment (milljoniaer) [reply src/components/pages/career-details/new-career-form/upload.ts]: I changed the name and I will do some testing after implementing the changes 👍

File: src/components/pages/career-details/new-career-form/career-form-fields.tsx
Comment (milljoniaer) [reply src/components/pages/career-details/new-career-form/career-form-fields.tsx:75]: I am ambiguous about this.
It is still a large component in career-form, since some of the props have still to be passed. 
But it is shorter and that is helpful

File: src/components/pages/career-details/new-career-form/career-form.tsx
Comment (milljoniaer) [reply src/components/pages/career-details/new-career-form/career-form.tsx:84]: This validation of the selectedFiles is generally something i am not happy with. I am already thinking of a way to put this whole thing into the file dropper.

But for now, I would suggest to leave it like this, and I will plan to refactor the file dropper some day to a controlled component, I have an idea in which way it could work, but it is not implementation ready

File: src/components/pages/career-details/new-career-form/upload.ts
Comment (milljoniaer) [reply src/components/pages/career-details/new-career-form/upload.ts]: Test Application has worked with all fields and an custom channel 👍

**Pull request merged**
